### PR TITLE
Custom Update UI

### PR DESCRIPTION
**Pull request by JoshuaBrest**

### CODE REVIEW

File: Whisky/Views/Updater/UpdatePreviewView.swift [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    func getChangelog() async {
+        withAnimation { markdownTextState = .loading }
+        let ghOwner = (Bundle.main.object(forInfoDictionaryKey: "GithubRepoOwner") as? String) ?? "Whisky-App"
+        let ghRepo = (Bundle.main.object(forInfoDictionaryKey: "GithubRepoName") as? String) ?? "Whisky"
+
+        // Make a request to the Github API to get the latest release
+        // Append path not using string interpolation to non-urlencoded paths
+        guard let url = URL(string: "https://api.github.com/")?
```
Comment (IsaacMarovitz) [inline Whisky/Views/Updater/UpdatePreviewView.swift:113]: Can this not use the current appcast? We can't connect to GitHub as it's blocked in China.

File: Whisky/Views/Updater/UpdatePreviewView.swift [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            VStack {
+                ScrollView {
+                    VStack(alignment: .leading) {
+                        Text("update.changeLog")
+                            .font(.title2)
+                            .fontWeight(.bold)
+                            .padding(.bottom, 12)
+                    }
+                    .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .leading)
+                    Markdown {
```
Comment (IsaacMarovitz) [inline Whisky/Views/Updater/UpdatePreviewView.swift:74]: SwiftUI Text has native support for Markdown: https://www.hackingwithswift.com/quick-start/swiftui/how-to-render-markdown-content-in-text

File: Whisky/Info.plist [hunk 3]
```diff
@@ -57,6 +57,10 @@
 	<string>https://data.getwhisky.app/appcast.xml</string>
 	<key>SUPublicEDKey</key>
 	<string>tnZFAvPUfCpM7Tr7Sx5gKRm6BUQQ6htJQeOMP44evms=</string>
+	<key>GithubRepoOwner</key>
```
Comment (IsaacMarovitz) [inline Whisky/Info.plist:60]: These keys are unused

File: Whisky/Views/Updater/UpdatePreviewView.swift (hunk 1 above)
Comment (JoshuaBrest) [reply Whisky/Views/Updater/UpdatePreviewView.swift]: Okay, it's now using the xml file as requested

File: Whisky/Views/Updater/UpdatePreviewView.swift [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+//  You should have received a copy of the GNU General Public License along with Whisky.
+//  If not, see https://www.gnu.org/licenses/.
+//
+
+import SwiftUI
+import Sparkle
+import MarkdownUI
+
+struct UpdatePreviewView: View {
+//    enum MarkdownTextState {
```
Comment (IsaacMarovitz) [inline Whisky/Views/Updater/UpdatePreviewView.swift:24]: Please cleanup of debug values

File: Whisky/Views/Updater/UpdatePreviewView.swift [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+                    .markdownTheme(.basic)
+                }
+                .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .leading)
+            }
+            .frame(maxWidth: .infinity, maxHeight: .infinity)
+            .padding(20)
+            .background(.ultraThickMaterial)
+        }
+        .frame(maxWidth: .infinity, maxHeight: .infinity)
+        .onAppear {
```
Comment (IsaacMarovitz) [inline Whisky/Views/Updater/UpdatePreviewView.swift:86]: More cleanup needed

**Review by IsaacMarovitz — CHANGES_REQUESTED** [review]

File: Whisky/Views/Updater/UpdatePreviewView.swift (hunk 2 above)
Comment (JoshuaBrest) [reply Whisky/Views/Updater/UpdatePreviewView.swift:74]: Yes but not it doesn't support images and code blocks

File: Whisky/Views/Updater/UpdatePreviewView.swift (hunk 5 above)
Comment (JoshuaBrest) [reply Whisky/Views/Updater/UpdatePreviewView.swift:86]: Mb I forgot to save the file

File: Whisky/Views/Updater/UpdatePreviewView.swift (hunk 2 above)
Comment (IsaacMarovitz) [reply Whisky/Views/Updater/UpdatePreviewView.swift:74]: We don't need images or code blocks

File: Whisky/Views/Updater/UpdatePreviewView.swift [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+import SwiftUI
+import Sparkle
+import MarkdownUI
+
+struct UpdatePreviewView: View {
+    let dismiss: () -> Void
+    let install: () -> Void
+    @Binding var markdownText: String
+    @Binding var nextVersion: String
```
Comment (cuba) [inline Whisky/Views/Updater/UpdatePreviewView.swift:27]: I don't see any reason why these need to be `@Bindng` a simple let should work just fine here.

File: Whisky/Utils/SparkleUpdaterEvents.swift [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+    var checkingForUpdates: ((@escaping () -> Void) -> Void)?
+    var updateFound: ((SUAppcastItem, SPUUserUpdateState, @escaping (SPUUserUpdateChoice) -> Void) -> Void)?
+    var update: ((SPUDownloadData) -> Void)?
+    var updateError: ((NSError) -> Void)?
+    var updateDownloadState: (() -> Void)?
+    var updateExtractState: (() -> Void)?
+    var updateInstallState: (() -> Void)?
+    var updateReadyRelaunch: ((@escaping (SPUUserUpdateChoice) -> Void) -> Void)?
+    var updateDismiss: (() -> Void)?
```
Comment (cuba) [inline Whisky/Utils/SparkleUpdaterEvents.swift:33]: I think we can use `ObservedObject` here and avoid all these callbacks

File: Whisky/Views/Updater/UpdatePreviewView.swift (hunk 6 above)
Comment (JoshuaBrest) [reply Whisky/Views/Updater/UpdatePreviewView.swift:27]: I tried that yet somehow it didn't work.

File: Whisky/Views/Updater/UpdatePreviewView.swift [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+//  You should have received a copy of the GNU General Public License along with Whisky.
+//  If not, see https://www.gnu.org/licenses/.
+//
+
+import SwiftUI
+import Sparkle
+import MarkdownUI
+
+struct UpdatePreviewView: View {
+    let dismiss: () -> Void
```
[the remark this answers is not in the document]
Comment (JoshuaBrest) [reply Whisky/Views/Updater/UpdatePreviewView.swift:24]: I don'r see how that would work, we would still need to tell sparkle that the popup was dismissed.

File: Whisky/Info.plist (hunk 3 above)
Comment (JoshuaBrest) [reply Whisky/Info.plist]: They are gone now.

File: Whisky/Views/Updater/UpdatePreviewView.swift (hunk 4 above)
Comment (JoshuaBrest) [reply Whisky/Views/Updater/UpdatePreviewView.swift]: Done

**Pull request closed without merging**